lib/request-execution.js

```diff
@@ -0,0 +1,231 @@
+  this._request = this._parent.request;
+}
+
+/**
+ * Starts the execution by borrowing the next connection available using the query plan.
+ * It invokes the callback when a connection is acquired, if any.
+ * @param {Function} getHostCallback
+ */
+RequestExecution.prototype.start = function (getHostCallback) {
+  var self = this;
+  getHostCallback = getHostCallback || utils.noop;
+  this._parent.getNextConnection(function nextConnectionCallback(err, connection, host) {
+    if (err) {
+      return self._parent.handleNoHostAvailable(err, self);
+    }
+    self._connection = connection;
+    self._host = host;
+    getHostCallback(host);
+    if (self._cancelled) {
+      // No need to send the request
+      return;
+    }
+    self._parent.speculativeExecutions++;
+    self._sendOnConnection();
+  });
+};
+
+RequestExecution.prototype._sendOnConnection = function () {
+  var self = this;
+  this._operation =
+    this._connection.sendStream(this._request, this._parent.requestOptions, function responseCb(err, response) {
+      if (self._cancelled) {
+        // Avoid handling the response / err
+        return;
+      }
+      if (err) {
+        return self._handleError(err);
+      }
+      var result = self._getResultSet(response);
+      if (response.schemaChange) {
+        return self._parent.client.handleSchemaAgreementAndRefresh(
+          self._connection, response.schemaChange, function schemaCb(){
+            self._parent.setCompleted(null, result);
+          });
+      }
+      if (response.keyspaceSet) {
+        self._parent.client.keyspace = response.keyspaceSet;
+      }
+      self._parent.setCompleted(null, result);
+    });
+};
+
+RequestExecution.prototype._getResultSet = function (response) {
+  return new types.ResultSet(response, this._host.address, this._parent.triedHosts, this._parent.speculativeExecutions, this._request.consistency);
+};
+
+/**
+ * Allows the handler to cancel the current request.
+ * When the request has been already written, we can unset the callback and forget about it.
+ */
+RequestExecution.prototype.cancel = function () {
+  this._cancelled = true;
+  if (this._operation === null) {
+    return;
+  }
+  this._operation.cancel();
+};
+
+RequestExecution.prototype._handleError = function (err) {
+  this._parent.triedHosts[this._host.address] = err;
+  err['coordinator'] = this._host.address;
+  if (err.code === types.responseErrorCodes.unprepared && (err instanceof errors.ResponseError)) {
+    return this._prepareAndRetry(err.queryId);
+  }
+  var decisionInfo = this._getDecision(err);
+  if (err.isSocketError) {
+    this._host.removeFromPool(this._connection);
+  }
+  if (!decisionInfo || decisionInfo.decision === retry.RetryPolicy.retryDecision.rethrow) {
+    if (this._request instanceof requests.QueryRequest || this._request instanceof requests.ExecuteRequest) {
+      err['query'] = this._request.query;
+    }
+    return this._parent.setCompleted(err);
```

Comment:
`this._operation.setResult(err)` should be used instead, as we need to check that the operation hasn't been already cancelled.

There should be only one call to `this._parent.setCompleted()` and that is on `this._sendOnConnection().

Reply:
Don't mind, the correct flow is:
- `Connection` should set the `OperationState` as completed or timed out.
- `RequestExecution` should signal the parent `RequestHandler` as completed.

Reply:
This seems to be working great so far.  Have been running test for about 8 hours without issues with a 1.65% speculative execution rate.